src/library/blast/arith/heuristic.cpp

```diff
@@ -0,0 +1,552 @@
+        blast_tmp_type_context m_tmp_tctx;
+        bool ok = static_cast<bool>(m_tmp_tctx->mk_class_instance(get_app_builder().mk_linear_ordered_comm_ring(A)));
+        if (!ok) lean_trace(*g_heuristic_trace_name, tout() << "bad type: " << ppb(A) << "\n";);
+        return ok;
+    }
+    poly linearize(expr const & A, expr const & rhs, bool strict, lazy_proof const & lproof) {
+        polynomial p = arith::normalize_poly(rhs);
+        return poly(A, to_list(p.get_monomials().begin(), p.get_monomials().end()), p.get_offset(), strict, lproof);
+    }
+public:
+    list<poly> linearize(hypothesis_idx hidx) {
+        hypothesis const & h = curr_state().get_hypothesis_decl(hidx);
+        /* TODO(dhs): as a pre-process step, put 0 on the LHS of the inequality. */
+        expr A, lhs, rhs;
+        if (is_lt(h.get_type(), A, lhs, rhs) && type_ok(A)) {
+            expr new_rhs = get_app_builder().mk_add(A, rhs, get_app_builder().mk_neg(A, lhs));
+            return list<poly>(linearize(A, new_rhs, true, [=]() {
+                        return get_app_builder().mk_app(get_ordered_arith_zero_lt_of_lt_name(), h.get_self());
+                    }));
+        } else if (is_le(h.get_type(), A, lhs, rhs) && type_ok(A)) {
+            expr new_rhs = get_app_builder().mk_add(A, rhs, get_app_builder().mk_neg(A, lhs));
+            return list<poly>(linearize(A, new_rhs, false, [=]() {
+                        return get_app_builder().mk_app(get_ordered_arith_zero_le_of_le_name(), h.get_self());
+                    }));
+        } else if (is_eq(h.get_type(), A, lhs, rhs) && type_ok(A)) {
+            expr new_rhs1 = get_app_builder().mk_add(A, rhs, get_app_builder().mk_neg(A, lhs));
+            expr new_rhs2 = get_app_builder().mk_add(A, lhs, get_app_builder().mk_neg(A, rhs));
+            poly p1 = linearize(A, new_rhs1, false, [=]() {
+                    return get_app_builder().mk_app(get_ordered_arith_zero_le_of_eq1_name(), h.get_self());
+                });
+            poly p2 = linearize(A, new_rhs2, false, [=]() {
+                    return get_app_builder().mk_app(get_ordered_arith_zero_le_of_eq2_name(), h.get_self());
+                });
+            return list<poly>({p1, p2});
+        } else {
+            return list<poly>();
+        }
+    }
+};
+
+/* Poly-pot */
+class poly_pot {
+    list<poly> m_positives;
+    list<poly> m_negatives;
+public:
+    void insert(poly const & p) {
+        if (p.is_positive()) {
+            m_positives = cons(p, m_positives);
+        } else {
+            m_negatives = cons(p, m_negatives);
+        }
+    }
+    list<poly> get_positives() { return m_positives; }
+    list<poly> get_negatives() { return m_negatives; }
+};
+
+struct heuristic_branch_extension : public branch_extension {
+private:
+    // TODO(dhs): will this cause problems during rebalancing?
+    // Should I wrap this class and make a cheap copy?
+    typedef rb_map<std::vector<atom>, poly_pot, atoms_quick_cmp> poly_pot_map;
```

Comment:
You have been using lists. Why did you use a vector here? The map performs copies during insertion.